### Data Cleanup
In this notebook we look at the data provided and see if anything is missing, have nulls values, outliers etc ... and then do some cleanup to make sure the data is in a usable state for further analysis

##### [NOTE:]The comments on the analysis is documented after the each cell code

In [114]:
# import all required libraries for data cleanup
import pandas as pd
import numpy as np

In [115]:
# load the coupons.csv data using read_csv method in pandas

data = pd.read_csv('data/coupons.csv')

In [116]:
#inspect to see whether the dataframe has any data or not
data.head()

destination  passanger weather  temperature  time  \
0  No Urgent Place      Alone   Sunny           55   2PM   
1  No Urgent Place  Friend(s)   Sunny           80  10AM   
2  No Urgent Place  Friend(s)   Sunny           80  10AM   
3  No Urgent Place  Friend(s)   Sunny           80   2PM   
4  No Urgent Place  Friend(s)   Sunny           80   2PM   

                  coupon expiration  gender age      maritalStatus  ...  \
0        Restaurant(<20)         1d  Female  21  Unmarried partner  ...   
1           Coffee House         2h  Female  21  Unmarried partner  ...   
2  Carry out & Take away         2h  Female  21  Unmarried partner  ...   
3           Coffee House         2h  Female  21  Unmarried partner  ...   
4           Coffee House         1d  Female  21  Unmarried partner  ...   

   CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
0        never       NaN                  4~8              1~3   
1        never       NaN                  4~8              1~3   
2        never       NaN                  4~8              1~3   
3        never       NaN                  4~8              1~3   
4        never       NaN                  4~8              1~3   

  toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
0                1                 0                 0              0   
1                1                 0                 0              0   
2                1                 1                 0              0   
3                1                 1                 0              0   
4                1                 1                 0              0   

  direction_opp  Y  
0             1  1  
1             1  0  
2             1  1  
3             1  0  
4             1  0  

[5 rows x 26 columns]

In [117]:
# check how many non-null counts for each feature
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12684 entries, 0 to 12683
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   destination           12684 non-null  object
 1   passanger             12684 non-null  object
 2   weather               12684 non-null  object
 3   temperature           12684 non-null  int64 
 4   time                  12684 non-null  object
 5   coupon                12684 non-null  object
 6   expiration            12684 non-null  object
 7   gender                12684 non-null  object
 8   age                   12684 non-null  object
 9   maritalStatus         12684 non-null  object
 10  has_children          12684 non-null  int64 
 11  education             12684 non-null  object
 12  occupation            12684 non-null  object
 13  income                12684 non-null  object
 14  car                   108 non-null    object
 15  Bar                   12577 non-null

In [118]:
# find out the number of null values for each given column
data.isna().sum()

destination                 0
passanger                   0
weather                     0
temperature                 0
time                        0
coupon                      0
expiration                  0
gender                      0
age                         0
maritalStatus               0
has_children                0
education                   0
occupation                  0
income                      0
car                     12576
Bar                       107
CoffeeHouse               217
CarryAway                 151
RestaurantLessThan20      130
Restaurant20To50          189
toCoupon_GEQ5min            0
toCoupon_GEQ15min           0
toCoupon_GEQ25min           0
direction_same              0
direction_opp               0
Y                           0
dtype: int64

In above sum of null values, are many columns has nulls value, to name them car,bar,CoffeeHouse, CarryAway,RestaurantLessThan20 and Restaurant20To50.

Some of these features(columns) are very important to answer the question of 'will customer accepts the coupon' for example bar,CoffeeHouse, CarryAway,RestaurantLessThan20 and Restaurant20To50.

So we may have to replace these nulls with maybe a median of the data if the column is of numerical data else we may have to drop the column data.


In [119]:
# Lets drop the null values
data.dropna()

destination  passanger weather  temperature  time  \
1237  No Urgent Place      Alone   Sunny           55   2PM   
1238  No Urgent Place  Friend(s)   Sunny           80  10AM   
1239  No Urgent Place  Friend(s)   Sunny           80  10AM   
1240  No Urgent Place  Friend(s)   Sunny           80  10AM   
1241  No Urgent Place  Friend(s)   Sunny           80   2PM   
...               ...        ...     ...          ...   ...   
8293             Work      Alone   Rainy           55   7AM   
8294             Work      Alone   Rainy           55   7AM   
8295             Work      Alone   Snowy           30   7AM   
8296             Work      Alone   Snowy           30   7AM   
8297             Work      Alone   Sunny           80   7AM   

                     coupon expiration gender age maritalStatus  ...  \
1237        Restaurant(<20)         1d   Male  26        Single  ...   
1238           Coffee House         2h   Male  26        Single  ...   
1239                    Bar         1d   Male  26        Single  ...   
1240  Carry out & Take away         2h   Male  26        Single  ...   
1241           Coffee House         1d   Male  26        Single  ...   
...                     ...        ...    ...  ..           ...  ...   
8293        Restaurant(<20)         2h   Male  21        Single  ...   
8294      Restaurant(20-50)         1d   Male  21        Single  ...   
8295        Restaurant(<20)         2h   Male  21        Single  ...   
8296      Restaurant(20-50)         1d   Male  21        Single  ...   
8297        Restaurant(<20)         1d   Male  21        Single  ...   

      CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
1237          4~8       4~8                  1~3            never   
1238          4~8       4~8                  1~3            never   
1239          4~8       4~8                  1~3            never   
1240          4~8       4~8                  1~3            never   
1241          4~8       4~8                  1~3            never   
...           ...       ...                  ...              ...   
8293        less1       1~3                less1            less1   
8294        less1       1~3                less1            less1   
8295        less1       1~3                less1            less1   
8296        less1       1~3                less1            less1   
8297        less1       1~3                less1            less1   

     toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
1237                1                 0                 0              0   
1238                1                 0                 0              0   
1239                1                 0                 0              0   
1240                1                 1                 0              0   
1241                1                 0                 0              0   
...               ...               ...               ...            ...   
8293                1                 1                 1              0   
8294                1                 1                 1              0   
8295                1                 1                 1              0   
8296                1                 1                 1              0   
8297                1                 0                 0              1   

     direction_opp  Y  
1237             1  1  
1238             1  0  
1239             1  1  
1240             1  0  
1241             1  1  
...            ... ..  
8293             1  1  
8294             1  1  
8295             1  1  
8296             1  1  
8297             0  1  

[108 rows x 26 columns]

The above dropna() mwthod will drop most of the rows as 'car'has more rows with null values.
The resultant dataframe has only 108 which far less than original dataset which 12684, so dropping null value columns is not a good idea.

In [120]:
# Lets check how categorical the 'car' column is and what values it has
data['car'].value_counts()

Scooter and motorcycle                      22
Mazda5                                      22
do not drive                                22
crossover                                   21
Car that is too old to install Onstar :D    21
Name: car, dtype: int64

When we look at the 'car' column values i dont think it will add any value for this analysis so either we can drop just the column of leave it as is and dont use this column for analysis purpose.

For the purpose of this study im going to drop the column for now and see how it impacts the study.

In [121]:
# dropping the column 'car'
data.drop('car', axis=1, inplace=True)

In [122]:
data.isna().sum()

destination               0
passanger                 0
weather                   0
temperature               0
time                      0
coupon                    0
expiration                0
gender                    0
age                       0
maritalStatus             0
has_children              0
education                 0
occupation                0
income                    0
Bar                     107
CoffeeHouse             217
CarryAway               151
RestaurantLessThan20    130
Restaurant20To50        189
toCoupon_GEQ5min          0
toCoupon_GEQ15min         0
toCoupon_GEQ25min         0
direction_same            0
direction_opp             0
Y                         0
dtype: int64

In [123]:
data['Bar'].value_counts()

never    5197
less1    3482
1~3      2473
4~8      1076
gt8       349
Name: Bar, dtype: int64

In [124]:
data['CoffeeHouse'].value_counts()

less1    3385
1~3      3225
never    2962
4~8      1784
gt8      1111
Name: CoffeeHouse, dtype: int64

In [125]:
data['CarryAway'].value_counts()

1~3      4672
4~8      4258
less1    1856
gt8      1594
never     153
Name: CarryAway, dtype: int64

In [126]:
data['RestaurantLessThan20'].value_counts()

1~3      5376
4~8      3580
less1    2093
gt8      1285
never     220
Name: RestaurantLessThan20, dtype: int64

In [127]:
data['Restaurant20To50'].value_counts()

less1    6077
1~3      3290
never    2136
4~8       728
gt8       264
Name: Restaurant20To50, dtype: int64

After looking at the values of the above columns, the values **less1/never** goes with value as **'0'** as these are visits and can only be whole number. So we can replace **'less1'** and **'never'** with **'0/never'** similar to other categorical values.

Updating these column values as **'never'**

In [128]:
# replace the all column values of less1 to never
data['Bar'].mask(data['Bar'] == 'less1','never', inplace=True)
data['CoffeeHouse'].mask(data['CoffeeHouse'] == 'less1','never', inplace=True)
data['CarryAway'].mask(data['CarryAway'] == 'less1','never', inplace=True)
data['RestaurantLessThan20'].mask(data['RestaurantLessThan20'] == 'less1','never', inplace=True)
data['Restaurant20To50'].mask(data['Restaurant20To50'] == 'less1','never', inplace=True)

In [129]:
# lets check to make sure the values are updated
data['CoffeeHouse'].value_counts()

never    6347
1~3      3225
4~8      1784
gt8      1111
Name: CoffeeHouse, dtype: int64

### Fill null values
The null value columns are bar,CoffeeHouse, CarryAway,RestaurantLessThan20 and Restaurant20To50. We can fill these columns with 'never' as the null values are very less and 'never' is the value of the most of the rows.

In [130]:
# fill all the above mentioned column values to 'never'
data.fillna('never', inplace=True)

In [131]:
# now no null values in the data set
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12684 entries, 0 to 12683
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   destination           12684 non-null  object
 1   passanger             12684 non-null  object
 2   weather               12684 non-null  object
 3   temperature           12684 non-null  int64 
 4   time                  12684 non-null  object
 5   coupon                12684 non-null  object
 6   expiration            12684 non-null  object
 7   gender                12684 non-null  object
 8   age                   12684 non-null  object
 9   maritalStatus         12684 non-null  object
 10  has_children          12684 non-null  int64 
 11  education             12684 non-null  object
 12  occupation            12684 non-null  object
 13  income                12684 non-null  object
 14  Bar                   12684 non-null  object
 15  CoffeeHouse           12684 non-null

### Column Names
Lets look at the column names and see if they are makes sense or not. If not rename them to easy to understand.

The last column name is **'Y'** which stands for whether the coupon is accepted or not. Lets rename this to **'coupon_accepted'** to easy to read.

In [132]:
# rename the column 'Y'
data.rename({'Y':'coupon_accepted'}, axis='columns', inplace=True)

In [133]:
# check to make sure whether its renamed or not
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12684 entries, 0 to 12683
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   destination           12684 non-null  object
 1   passanger             12684 non-null  object
 2   weather               12684 non-null  object
 3   temperature           12684 non-null  int64 
 4   time                  12684 non-null  object
 5   coupon                12684 non-null  object
 6   expiration            12684 non-null  object
 7   gender                12684 non-null  object
 8   age                   12684 non-null  object
 9   maritalStatus         12684 non-null  object
 10  has_children          12684 non-null  int64 
 11  education             12684 non-null  object
 12  occupation            12684 non-null  object
 13  income                12684 non-null  object
 14  Bar                   12684 non-null  object
 15  CoffeeHouse           12684 non-null

In [134]:
#Find the dupliates of the dataset
data[data.duplicated()]

destination passanger weather  temperature  time                 coupon  \
4192        Work     Alone   Sunny           80   7AM  Carry out & Take away   
4236        Work     Alone   Sunny           80   7AM  Carry out & Take away   
4280        Work     Alone   Sunny           80   7AM  Carry out & Take away   
4324        Work     Alone   Sunny           80   7AM  Carry out & Take away   
4409        Work     Alone   Sunny           80   7AM  Carry out & Take away   
...          ...       ...     ...          ...   ...                    ...   
8511        Home     Alone   Sunny           80   6PM                    Bar   
8512        Home   Partner   Sunny           30  10PM  Carry out & Take away   
8513        Work     Alone   Rainy           55   7AM        Restaurant(<20)   
8515        Work     Alone   Snowy           30   7AM      Restaurant(20-50)   
8516        Work     Alone   Sunny           80   7AM        Restaurant(<20)   

     expiration  gender age    maritalStatus  ...  CoffeeHouse CarryAway  \
4192         1d    Male  26           Single  ...        never       1~3   
4236         1d    Male  26           Single  ...          gt8       gt8   
4280         1d  Female  26           Single  ...        never       4~8   
4324         1d  Female  46           Single  ...        never       4~8   
4409         1d  Female  21           Single  ...        never     never   
...         ...     ...  ..              ...  ...          ...       ...   
8511         2h    Male  46  Married partner  ...          1~3       1~3   
8512         2h    Male  46  Married partner  ...          1~3       1~3   
8513         2h    Male  46  Married partner  ...          1~3       1~3   
8515         1d    Male  46  Married partner  ...          1~3       1~3   
8516         1d    Male  46  Married partner  ...          1~3       1~3   

     RestaurantLessThan20 Restaurant20To50 toCoupon_GEQ5min toCoupon_GEQ15min  \
4192                never            never                1                 1   
4236                  4~8            never                1                 1   
4280                  1~3            never                1                 1   
4324                  1~3              1~3                1                 1   
4409                  1~3            never                1                 1   
...                   ...              ...              ...               ...   
8511                never              1~3                1                 0   
8512                never              1~3                1                 1   
8513                never              1~3                1                 1   
8515                never              1~3                1                 1   
8516                never              1~3                1                 0   

     toCoupon_GEQ25min direction_same direction_opp  coupon_accepted  
4192                 1              0             1                1  
4236                 1              0             1                1  
4280                 1              0             1                1  
4324                 1              0             1                1  
4409                 1              0             1                0  
...                ...            ...           ...              ...  
8511                 0              1             0                1  
8512                 0              0             1                1  
8513                 1              0             1                0  
8515                 1              0             1                0  
8516                 0              1             0                1  

[88 rows x 25 columns]

In [138]:
#Remove duplicate rows
data.drop_duplicates(keep='first', inplace=True)

In [139]:
# Save the processing data as 'coupons_processed.csv'
data.to_csv('data/coupons_processed.csv',index=False)